bundles/org.openhab.binding.mercedesme/src/main/java/org/openhab/binding/mercedesme/internal/api/Authorization.java

```diff
@@ -0,0 +1,533 @@
+        // restore token from persistence if available
+        String storedToken = storage.get(identifier);
+        if (storedToken != null) {
+            token = Utils.INVALID_TOKEN;
+            try {
+                TokenResponse tokenResponseJson = Utils.GSON.fromJson(storedToken, TokenResponse.class);
+                token = decodeToken(tokenResponseJson);
+            } catch (JsonSyntaxException e) {
+                logger.warn("Stored token {} for {} not parsable: {}", storedToken, config.email, e.getMessage());
+            }
+            if (!authTokenIsValid()) {
+                storage.remove(identifier);
+                logger.trace("Invalid token for {}", config.email);
+            } else {
+                atrl.onAccessTokenResponse(token);
+            }
+        } else {
+            logger.trace("No token for {} stored, stay on invalid token", config.email);
+        }
+    }
+
+    protected synchronized String getToken() {
+        if (token.isExpired(Instant.now(), EXPIRATION_BUFFER)) {
+            if (authTokenIsValid()) {
+                refreshToken();
+            }
+        }
+        return token.getAccessToken();
+    }
+
+    private void refreshToken() {
+        logger.trace("RefreshToken");
+        try {
+            String url = Utils.getTokenUrl(config.region);
+            Request req = httpClient.POST(url);
+            req.header("X-Device-Id", UUID.randomUUID().toString());
+            req.header("X-Request-Id", UUID.randomUUID().toString());
+
+            String grantAttribute = "grant_type=refresh_token";
+            String refreshTokenAttribute = "refresh_token="
+                    + URLEncoder.encode(token.getRefreshToken(), StandardCharsets.UTF_8.toString());
+            String content = grantAttribute + "&" + refreshTokenAttribute;
+            req.header(HttpHeader.CONTENT_TYPE, "application/x-www-form-urlencoded");
+            req.content(new StringContentProvider(content));
+
+            ContentResponse cr = req.timeout(Constants.REQUEST_TIMEOUT_MS, TimeUnit.MILLISECONDS).send();
+            int tokenResponseStatus = cr.getStatus();
+            String tokenResponse = cr.getContentAsString();
+            if (tokenResponseStatus == 200) {
+                storeToken(tokenResponse);
+            } else {
+                token = Utils.INVALID_TOKEN;
+                /**
+                 * 1) remove token from storage
+                 * 2) listener will be informed about INVALID_TOKEN and bridge will go OFFLINE
+                 */
+                storage.remove(identifier);
+                logger.warn("Failed to refresh token {} {}", tokenResponseStatus, tokenResponse);
+            }
+            listener.onAccessTokenResponse(token);
+        } catch (InterruptedException | TimeoutException | ExecutionException | UnsupportedEncodingException
+                | JsonSyntaxException e) {
+            logger.info("Failed to refresh token {}", e.getMessage());
```

Comment:
Should this be propagated to the user as a Thing status description?

Reply:
I wouldn't do at this location.  If a data call or websocket start fails with invalid token `AccountHandler `tries to recover the situation with a new login. If this fails then it's time to present `OFFLINE `status.

Reply:
I guess it will be offline until it successfully recovers then? It would be offline with text "Refresh token expired, attempting relogin" or similar. This would make it more visible for users not checking logs, especially in cases where it doesn't succeed. Anyway, this is minor and you can also choose to leave it as is.